lib/ansible/modules/cloud/docker/docker_node.py

```diff
@@ -0,0 +1,301 @@
+#!/usr/bin/python
+#
+# (c) 2019 Piotr Wojciechowski <piotr@it-playground.pl>
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+
+__metaclass__ = type
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+DOCUMENTATION = '''
+---
+module: docker_node
+short_description: Manage Docker Swarm node
+version_added: "2.8"
+description:
+    - Manages the Docker nodes via Swarm Manager.
+    - This module allows to change the node's role, its availability, and to modify, add or remove node labels.
+options:
+    hostname:
+        description:
+            - The hostname or ID of node as registered in Swarm.
+            - If more than one node is registered using the same hostname the ID must be used,
+              otherwise task will fail.
+        required: true
+        type: str
+    labels:
+        description: User-defined key/value metadata. If not provided then labels assigned to node remains unchanged.
+        type: dict
+    labels_state:
+        description:
+            - Defines the operation on the labels assigned to node.
+            - Set to C(merge) to combine labels provided in I(labels) with those already assigned to the node.
+              If no labels are assigned then it will add listed labels. The I(labels) must be specified.
+            - If set as C(merge) for labels that are assigned to node will update their values as specified in I(labels).
+            - If set as C(merge) labels assigned to node that are not specified in I(labels) will remain unchanged.
+            - If set as C(merge) will add new labels specified in I(labels) if labels are not assigned to node.
+            - Set to C(replace) to replace assigned labels with provided ones. The I(labels) must be specified.
+            - Set to C(remove) to remove labels specified in I(labels). If I(labels) is not provided then will remove
```

Comment:
The comparisons feature for labels is pretty simple: you can replace labels, update labels, or ignore labels.

In genral: Ansible is not about doing explicit operation, but about specifying state and letting Ansible make sure objects adhere to this state. That's why such explicit operation options are not OK in my opinion. Having an option to make sure some specific labels are not there is something I would accept, but not an interface which is different from the general Ansible way.

Reply:
So to get it all straight we are talking about implementation as below?

```
- name: Add labels (update the value if label exists)
  docker_node:
    hostname: mynode
    labels:
      key: value
    labels_state: merge

- name: Remove selected labels (if exists)
  docker_node:
    hostname: mynode
      labels_to_remove:
      - key1
      - key2

- name: Replace labels
  docker_node:
    hostname: mynode
    labels:
      key: value
    labels_state: replace

- name: Remove labels
  docker_node:
    hostname: mynode
    labels: {}
    labels_state: replace

- name: Add/update and remove at the same time
  docker_node:
    hostname: mynode
    labels:
      key: value
    labels_state: merge
    labels_to_remove:
      - key2
```

In the last example if both non-empty `labels` and `labels_to_remove` are present then if the label is set in both options it will be updated, not removed (it will be clearly stated in the documentation of course) unless the `labels` is `{ }` and `labels_state` is `replace`, in such case it will remove all labels.

Reply:
Yes, I think this is good. The `labels_to_remove` is kind of an exception; maybe it would also be better to call it `labels_that_should_not_be_there` instead, though that's a pretty horrible name ;) (If you can come up with something better, feel free...)

Reply:
Might be just `remove_labels` or in less grammatical form `labels_remove`